client-app/src/common/services/query-builder.service.ts

```diff
@@ -0,0 +1,34 @@
+import { SearchCriteria } from '../models/search/search-criteria';
+import { SearchQuery } from '../models/search/search-query';
+
+export class QueryBuilder<
+TSearchCriteria extends SearchCriteria,
+TSearchQuery extends SearchQuery> {
+
+  constructor(
+    private searchCriteriaType: { new(): TSearchCriteria },
+    private searchQueryType: { new(): TSearchQuery }) {
+  }
+
+  toQueryObject(searchCriteria: TSearchCriteria): object {
+    return searchCriteria.toSearchQuery(this.searchQueryType).normalize();
+  }
+
+  toQuery(searchCriteria: TSearchCriteria): URLSearchParams {
+    return new URLSearchParams(Object.entries(this.toQueryObject(searchCriteria)) as string[][])
+  }
+
+  fromQueryObject(queryObject: object): TSearchCriteria {
```

Comment:
what exactly this method should do in accordance with its name? Maybe we should rename it to "toSearchCriteria"?

Reply:
Fixed method names. I used `parseQuery` and `parseURLSearchParams`. I can't use something like `convertToSearchCriteria` because there are two methods which do similar job (one for query object, another for URLSearchParams), but typescript doesn't support method overloading